lib/Macros/Sources/SwiftMacros/DebugDescriptionMacro.swift

```diff
@@ -0,0 +1,287 @@
+}
+
+private func syntaxKeyword<Decl: DeclSyntaxProtocol>(for decl: Decl) -> String {
+    if decl.is(StructDeclSyntax.self) { return "struct" }
+    if decl.is(ClassDeclSyntax.self) { return "class" }
+    if decl.is(EnumDeclSyntax.self) { return "enum" }
+    if decl.is(ExtensionDeclSyntax.self) { return "extension" }
+    assertionFailure("expected struct/class/enum/extension")
+    return "declaration"
+}
+
+private func genericParameters<Decl: DeclSyntaxProtocol>(_ decl: Decl) -> GenericParameterListSyntax? {
+    decl.as(StructDeclSyntax.self)?.genericParameterClause?.parameters ??
+    decl.as(ClassDeclSyntax.self)?.genericParameterClause?.parameters ??
+    decl.as(EnumDeclSyntax.self)?.genericParameterClause?.parameters
+}
+
+private func bindingName(_ binding: PatternBindingSyntax) -> String? {
+    binding.pattern.as(IdentifierPatternSyntax.self)?.identifier.text
+}
+
+private func isComputedProperty(_ binding: PatternBindingSyntax) -> Bool {
+    guard let accessors = binding.accessorBlock?.accessors else {
+        // No accessor block, not computed.
+        return false
+    }
+
+    switch accessors {
+    case .accessors(let accessors):
+        return accessors.contains { $0.accessorSpecifier.tokenKind == .keyword(.get) }
+    case .getter:
+        return true
+    @unknown default:
+        return false
+    }
+}
+
+// Enumerate a MemberAccess expression to produce DeclReferences in left to right order.
+// `a.b.c` will result in callbacks for each DeclReference `a`, `b`, and then `c`.
+private func enumerateMembers(_ memberAccess: MemberAccessExprSyntax, _ action: (DeclReferenceExprSyntax) -> Void) {
+    if let baseMembers = memberAccess.base?.as(MemberAccessExprSyntax.self) {
+        enumerateMembers(baseMembers, action)
+    } else if let baseDecl = memberAccess.base?.as(DeclReferenceExprSyntax.self) {
+        action(baseDecl)
+    }
+    action(memberAccess.declName)
+}
+
+// MARK: - Encoding
+
+private func encodeTypeSummaryRecord(_ typeIdentifier: String, _ summaryString: String) -> String {
+    let encodedType = encodeString(typeIdentifier)
+    let encodedSummary = encodeString(summaryString)
+    let recordSize = UInt(encodedType.count + encodedSummary.count)
+    return """
+    /* version */ 1 as UInt8,
+    /* record size */ \(literalBytes(encodeULEB(recordSize))),
+    /* "\(typeIdentifier)" */ \(literalBytes(encodedType)),
+    /* "\(summaryString)" */ \(literalBytes(encodedSummary))
+    """
+}
+
+private func literalBytes(_ bytes: [UInt8]) -> String {
+    bytes.map({ "\($0) as UInt8" }).joined(separator: ", ")
+}
+
+private func encodeString(_ string: String) -> [UInt8] {
+    let size = UInt(string.utf8.count) + 1 // including null terminator
+    var bytes: [UInt8] = []
+    bytes.append(contentsOf: encodeULEB(size))
+    bytes.append(contentsOf: Array(string.utf8))
+    bytes.append(0) // null terminator
+    return bytes
+}
+
+private func encodeULEB(_ value: UInt) -> [UInt8] {
```

Comment:
I think it would be nice if all of these functions had at least one-line Docc comments.